### ISSUE TITLE
Bug: Failed to load plugin '@typescript-eslint' declared in '.eslintrc.js'

### ISSUE THREAD

**Issue by anonrig** [eslint/eslint#15672]
### Environment

Node version:  v17.6.0
npm version: 8.5.1
Local ESLint version: ~8.10.0
Global ESLint version: -
Operating System:  MacOS Monterey

### What parser are you using?

@typescript-eslint/parser

### What did you do?

- `react.js`

```javascript
module.exports = {
  root: true,
  env: {
    browser: true,
    es6: true,
    node: true,
  },
  extends: [
    './shared/prettier.js',
    './shared/core.js',
    './shared/typescript.js',
    './shared/react.js',
  ],
}
```

- `prettier.js`

```javascript
module.exports = {
  extends: ['prettier'],
  plugins: ['prettier'],
  rules: {
    'prettier/prettier': [
      'error',
      {
        arrowParens: 'always',
        bracketSameLine: false,
        printWidth: 100,
        quoteProps: 'consistent',
        semi: false,
        singleQuote: true,
        trailingComma: 'all',
      },
      { usePrettierrc: false },
    ],
  },
}
```

- `core.js`

```javascript
module.exports = {
  parser: '@typescript-eslint/parser',
  plugins: ['unused-imports', 'import'],
  extends: ['eslint:recommended'],
  rules: {
    'no-console': 1,
    'no-use-before-define': 'off',
    'unused-imports/no-unused-imports': 'error',
    'unused-imports/no-unused-vars': [
      'warn',
      {
        vars: 'all',
        varsIgnorePattern: '^_',
        args: 'after-used',
        argsIgnorePattern: '^_',
      },
    ],
    'import/order': [
      'warn',
      {
        'groups': [['builtin', 'external'], ['internal', 'parent', 'index'], 'sibling'],
        'pathGroups': [
          {
            pattern: '@/**',
            group: 'internal',
            position: 'before',
          },
        ],
        'alphabetize': { order: 'asc' },
        'newlines-between': 'ignore',
      },
    ],
    'import/prefer-default-export': 'off',
    'import/no-extraneous-dependencies': [
      'error',
      {
        devDependencies: true,
        optionalDependencies: true,
        peerDependencies: true,
      },
    ],
  },
}
```

- `typescript.js`

```javascript
module.exports = {
  extends: ['plugin:@typescript-eslint/recommended'],
  plugins: ['@typescript-eslint'],
  overrides: [
    {
      files: ['*.ts', '*.tsx', '*.d.ts'],
      parser: '@typescript-eslint/parser',
    },
  ],
  rules: {
    '@typescript-eslint/explicit-module-boundary-types': 'off',
    '@typescript-eslint/no-unused-vars': 'off',
    '@typescript-eslint/no-use-before-define': [
      'error',
      {
        variables: false,
      },
    ],
  },
}
```

- `react.js`

```javascript
module.exports = {
  extends: ['airbnb/hooks', 'plugin:react/recommended', 'plugin:react/jsx-runtime'],
  parser: '@typescript-eslint/parser',
  parserOptions: {
    ecmaFeatures: { jsx: true },
    ecmaVersion: 'latest',
    sourceType: 'module',
  },
  rules: {
    'react/require-default-props': 'off',
    'react/prop-types': 'off',
    'react/react-in-jsx-scope': 'off',
  },
  settings: {
    react: {
      version: 'detect',
    },
  },
}

```

### What did you expect to happen?

To work properly :)

### What actually happened?

```
Error: Failed to load plugin '@typescript-eslint' declared in '.eslintrc.js » /Users/yagiz/Developer/stack-monorepo/packages/utilities/base-config/linter/react.js » ./shared/typescript.js': Cannot find module '@typescript-eslint/eslint-plugin'
Require stack:
- /Users/yagiz/Developer/stack-monorepo/packages/stack/__placeholder__.js
Referenced from: /Users/yagiz/Developer/stack-monorepo/packages/utilities/base-config/linter/shared/typescript.js

Error: Failed to load plugin '@typescript-eslint' declared in '.eslintrc.js » /Users/yagiz/Developer/stack-monorepo/packages/utilities/base-config/linter/react.js » ./shared/typescript.js': Cannot find module '@typescript-eslint/eslint-plugin'
Require stack:
- /Users/yagiz/Developer/stack-monorepo/packages/stack/__placeholder__.js
Referenced from: /Users/yagiz/Developer/stack-monorepo/packages/utilities/base-config/linter/shared/typescript.js
    at Function.Module._resolveFilename (node:internal/modules/cjs/loader:933:15)
    at Function.resolve (node:internal/modules/cjs/helpers:108:19)
    at Object.resolve (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:2325:46)
    at ConfigArrayFactory._loadPlugin (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3376:33)
    at ConfigArrayFactory._loadExtendedPluginConfig (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3196:29)
    at ConfigArrayFactory._loadExtends (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3117:29)
    at ConfigArrayFactory._normalizeObjectConfigDataBody (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3058:25)
    at _normalizeObjectConfigDataBody.next (<anonymous>)
    at ConfigArrayFactory._normalizeObjectConfigData (/Users/yagiz/Developer/stack-monorepo/common/temp/node_modules/.pnpm/@eslint+eslintrc@1.2.0/node_modules/@eslint/eslintrc/dist/eslintrc.cjs:3003:20)
    at _normalizeObjectConfigData.next (<anonymous>)
Process finished with exit code -1
```

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by nzakas** [comment]
For help with the typescript plugin, please contact the typescript-eslint project first.

**Comment by anonrig** [comment]
@nzakas This is not a problem related to typescript-eslint. Any subdependency of a shared config (used inside plugin), causes issues, if you don't use `@rushstack/eslint-patch/modern-module-resolution`.
